### ISSUE TITLE
 [Feature Request] Add ability to generate union type resolver based on __typename field

### ISSUE THREAD

**Issue by gdhordain** [kobylynskyi/graphql-java-codegen#626]
**Is your feature request related to a problem? Please describe.**
When using union type with graphql-java-codegen, we need to define a way to tell Jackson how to choose which implementation depending on the data.
This step could be generated.

**Describe the solution you'd like**
The spec (https://graphql.org/learn/schema/#union-types) defines the `__typename` field which will be populated with the data type.
It would be possible to generate Jackson annotations & resolver based on the `__typename` field.
This behaviour could be linked to a configuration toggle, and could be overwritten by `customAnnotationsMapping` configuration.

**Describe alternatives you've considered**
Stay with current implementation.

**Comment by jxnu-liguobin** [comment]
If I understand it well, do you want to generate these annotations directly to the interface/class?
```
customAnnotationsMapping := {
  val mapping = new util.HashMap[String, util.List[String]]
  val annotations = new util.ArrayList[String]()
  annotations.add("@com.fasterxml.jackson.annotation.JsonTypeInfo(use=com.fasterxml.jackson.annotation.JsonTypeInfo.Id.NAME, include=com.fasterxml.jackson.annotation.JsonTypeInfo.As.PROPERTY,property = \"__typename\")")
  annotations.add(
    """@com.fasterxml.jackson.annotation.JsonSubTypes(value = Array(
      |        new com.fasterxml.jackson.annotation.JsonSubTypes.Type(value = classOf[HumanDO], name = "Human"),
      |        new com.fasterxml.jackson.annotation.JsonSubTypes.Type(value = classOf[DroidDO], name = "Droid")))""".stripMargin)
  mapping.put("Character", annotations)
  mapping.put("Droid.appearsIn", util.Arrays.asList("@com.fasterxml.jackson.module.scala.JsonScalaEnumeration(classOf[io.github.dreamylost.model.EpisodeDOTypeRefer])"))
  mapping.put("Human.appearsIn", util.Arrays.asList("@com.fasterxml.jackson.module.scala.JsonScalaEnumeration(classOf[io.github.dreamylost.model.EpisodeDOTypeRefer])"))
  mapping
}
```
This demand is very common, and I have considered it. The annotation is not generated actively because there may be many serialization ways in Java. So, currently, the library only generates the tool class for serialization of Scala enumeration types(jackson).

Considering that Jackson dependency has been introduced into the library, I think it's worth to implement it, but it may need a switch, maybe not everyone needs it.

**Comment by gdhordain** [comment]
Yes. One possibility is to generate the correct annotations on the union interface linking to the union member classes.

Another possibility - my first though, maybe unnecessarily complex - would be to generate a generic Jackson resolver in the model package:
```
package io.github.kobylynskyi.order.graphql.model;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.DatabindContext;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase;

public class GraphqlTypenameResolver extends TypeIdResolverBase {

    private JavaType superType;

    @Override
    public void init(JavaType baseType) {
        superType = baseType;
    }

    @Override
    public JavaType typeFromId(DatabindContext context, String typename) {
        try {
            var clazz = Class.forName("io.github.kobylynskyi.order.graphql.model." + typename);
            return context.constructSpecializedType(superType, clazz);
        } catch (ClassNotFoundException e) {
            System.err.println(e.getMessage());
            return null;
        }
    }

    @Override
    public JsonTypeInfo.Id getMechanism() {
        return JsonTypeInfo.Id.NAME;
    }

    @Override
    public String idFromValue(Object obj) {
        return idFromValueAndType(obj, obj.getClass());
    }

    @Override
    public String idFromValueAndType(Object obj, Class<?> subType) {
        return subType.getSimpleName();
    }
}
```

And annotate the Union interface with the following:
```
package io.github.kobylynskyi.order.graphql.model;

@javax.annotation.Generated(
    value = "com.kobylynskyi.graphql.codegen.GraphQLCodegen",
    date = "2020-12-31T23:59:59-0500"
)
@com.fasterxml.jackson.annotation.JsonTypeInfo(use = com.fasterxml.jackson.annotation.JsonTypeInfo.Id.NAME, property = "__typename")
@com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver(io.github.kobylynskyi.order.graphql.model.GraphqlTypenameResolver.class)
public interface Character {
}
```

---

In any case, I agree this behaviour would need a switch.
If enabled, we could still override the annotation with the `customAnnotationsMapping` property.

**Comment by jxnu-liguobin** [comment]
If @kobylynskyi accepts issue,I prefer the second one. 😸

**Comment by kobylynskyi** [comment]
@jxnu-liguobin @gdhordain the proposed solution sounds good to me.
Let's create another switch (something like "generateJacksonTypeIdResolver") with default value = false.
And if it is set to true then generate the TypeIdResolver class (proposed in the description) and add annotations to the union type.
One minor note: please consider that the modelPackageName can be null/empty.

**Comment by gdhordain** [comment]
In the case a modelPackageName is empty, I see that it defaults to packageName.
And if both modelPackageName & packageName, the class will be generated in the default root package.

Ok for the proposed switch name and default value.

I will send a PR soon.

**Comment by jxnu-liguobin** [comment]
@gdhordain It would be better if kotlin and scala could be provided. Otherwise it doesn't matter, I'll take care of it when I'm free. Because I'm not in a hurry at the moment. Our graphql interface and union types are very few.

**Comment by gdhordain** [comment]
I will have a try.

**Comment by kobylynskyi** [comment]
Released in v5.1.0 today :shipit:
